FileIOUtils.cpp

```diff
@@ -0,0 +1,426 @@
+                            ((fileformatcolor & 0xFF) << 16);
+                    QColor newcolor = QColor::fromRgb(color);
+                    newcolor.setAlpha(0xFF);
+                    SetColorFunction(selectedPalId, i + 4 * j, newcolor.rgba());
+                }
+            }
+            palfile.close();
+        }
+    }
+    else if (selectedfilter.compare("YY-CHR pal file (*.pal)") == 0)
+    {
+        QFile file(qFilePath);
+        file.open(QIODevice::ReadOnly);
+        int length;
+        if (!file.isOpen() || (length = (int) file.size()) < (3 * 16))
+        {
+            file.close();
+            QMessageBox::critical(parent, QString(QObject::tr("Error")), QString(QObject::tr("File size too small! It should be >= 48 bytes.")));
+            return false;
+        }
+
+        // Read data
+        unsigned char *paldata = new unsigned char[length];
+        file.read((char *) paldata, length);
+        file.close();
+        for(int j = 1; j < 16; ++j) // Skip the first color
+        {
+            int color = (paldata[3 * j] << 16) |
+                    (paldata[3 * j + 1] << 8) |
+                    paldata[3 * j + 2];
+            QColor newcolor = QColor::fromRgb(color);
+            newcolor.setAlpha(0xFF);
+            SetColorFunction(selectedPalId, j, newcolor.rgba());
+        }
+    }
+
+    else if (selectedfilter.compare("Raw Binary palette (*.bin)") == 0)
+    {
+         QByteArray tmppalettedata;
+         QFile palbinfile(qFilePath);
+         if(!palbinfile.open(QIODevice::ReadOnly))
+         {
+             QMessageBox::critical(parent, QString(QObject::tr("Error")), QString(QObject::tr("Cannot open file! \n").append(palbinfile.errorString())));
+             return false;
+         }
+         tmppalettedata = palbinfile.readAll();
+         if (palbinfile.size() != 32)
+         {
+             QMessageBox::critical(parent, QString(QObject::tr("Error")),
+                                   QString(QObject::tr("Internal error: File size isn't 32 bytes, current size: ")) +
+                                   QString::number(palbinfile.size()));
+             palbinfile.close();
+             return false;
+         }
+         palbinfile.close();
+
+         QVector<QRgb> tmppalette;
+         unsigned short *tmppaldata = new unsigned short[16];
+         memset(tmppaldata, 0, 32);
+         memcpy(tmppaldata, tmppalettedata.data(), qMin(32, tmppalettedata.size()));
+         ROMUtils::LoadPalette(&tmppalette, tmppaldata, true);
+         for (int i = 1; i < 16; i++) // Skip the first color
+         {
+             SetColorFunction(selectedPalId, i, tmppalette[i]);
+         }
+         delete[] tmppaldata;
+    }
+    return true;
+}
+
+/// <summary>
+/// Import tiles8x8 data from file to some editor instances.
+/// </summary>
+/// <param name="parent">
+/// the parent widget pointer used in model dialog.
+/// </param>
+/// <param name="ref_palette">
```

Comment:
Let's add validation to the beginning of the function that ref_palette (as a QVector) is size 16, or exits the function and displays an internal error message. This is because there is logic in the function which requires ref_palette to be size 16 or else there will be memory corruption

Reply:
done